File: kernels/integration/python/trapn_cc.py

import math
import time
import numpy as np

# Integrand
import psi


# Quadrature is Clenshaw-Curtis - trapezoidal rule from -oo to oo with n divisions
# https://en.wikipedia.org/wiki/Clenshaw%E2%80%93Curtis_quadrature

# Coordinate transformation
@np.vectorize
def transform_cc(t):
    L = 2.0
    return L/np.tan(t)

# Jacobian of the transformation
@np.vectorize
def jacobian_cc(t):
    L = 2.0
    return L/np.sin(t)**2

# https://stackoverflow.com/questions/46782444/how-to-convert-a-linear-index-to-subscripts-with-support-for-negative-strides
def ind2sub(idx, shape, indices):
    for i in range(len(shape)):
        s = idx % shape[i]
        idx -= s
        idx //= shape[i]
        indices[i] = s
    return indices

def get_npts():
    return (n-1)**ndim

# Assume the integrand goes to zero faster than 1/x^2, so the endpoints evaluate to zero.

def trapn_cc(ndim, n, fn):
    a = 0.0
    b = math.pi
    # Size of each interval
    h = (1.0/n)*(math.pi)
    hval = h**ndim

    total = 0.0
    nnm1 = [n-1]*ndim
    idx = np.zeros(ndim, dtype=np.int64)
    x = np.zeros(ndim)
    xt = np.zeros(ndim)

    npts = (n-1)**ndim
    for i in range(npts):
        idx = ind2sub(i, nnm1, idx) + 1
        x[:] = idx*h
        xt[:] = transform_cc(x)
        jac = np.prod(jacobian_cc(x))
        total += jac*fn(xt)

    return total*hval


def psi_fn(x):
    r1 = x[0:3]
    r2 = x[3:6]
    B = 0.1

    p = psi.psi(r1, r2, B)
    return p*p


if __name__ == '__main__':

    # Fixed for this integrand (2 particles in 3D space)
    ndim = 6

    # Number of points in each dimension
    n = 6

    npts = get_npts()
    print("npts: ", npts)
    start = time.perf_counter()
    val = trapn_cc(ndim, n, psi_fn)
    end = time.perf_counter()
    elapsed = end - start
    print('Integrated value: ',val)
    print('Elapsed time (s): ',elapsed,' rate (func evals/s): ',npts/elapsed)

File: kernels/integration/cpp/trapn_cc.cpp

#include <iostream>
#include <iomanip>
#include <cmath>
#include <omp.h>

const double pi = M_PI;

const int ndim = 6;

// Define the integrand

double mag(double r[3])
{
  return std::sqrt(r[0]*r[0] + r[1]*r[1] + r[2]*r[2]);
}

double orb(double R[3])
{
  double r = mag(R);
  double Z = 2.0;
  double norm = 0.5/std::sqrt(pi);
  return norm*std::exp(-Z*r);
}

double jastrow(double r12, double B)
{
  double A = 0.5;
  return std::exp(A*r12/(1.0 + B*r12) - A/B);
}

double psi(double r1[3], double r2[3], double B)
{
  double o1 = orb(r1);
  double o2 = orb(r2);

  double r12[3];
  r12[0] = r2[0] - r1[0];
  r12[1] = r2[1] - r1[1];
  r12[2] = r2[2] - r1[2];

  double r12_mag = mag(r12);
  double j = jastrow(r12_mag, B);

  return o1*o2*j;
}

double psi_fn(double xx[ndim])
{
  double r1[3];
  double r2[3];

  r1[0] = xx[0];
  r1[1] = xx[1];
  r1[2] = xx[2];

  r2[0] = xx[3];
  r2[1] = xx[4];
  r2[2] = xx[5];

  double B = 0.1;
  double p = psi(r1, r2, B);
  return p*p;
}

// https://stackoverflow.com/questions/46782444/how-to-convert-a-linear-index-to-subscripts-with-support-for-negative-strides
void ind2sub(int idx, const int* shape, int* indices)
{
    for (int i = 0; i < ndim; i++)
    {
        int s = idx % shape[i];
        idx -= s;
        idx /= shape[i];
        indices[i] = s;
    }
}

uint64_t get_npts(int ndim, int n)
{
    return std::pow((n-1), ndim);
}

double transform_cc(double t)
{
    double L = 2.0;
    return L/std::tan(t);
}

double jacobian_cc(double t)
{
    double L = 2.0;
    double s = std::sin(t);
    return L/(s*s);
}

double trapn_cc(int n)
{
    double h = pi/n;

    double total = 0.0;

    int npts = std::pow((n-1), ndim);
    int indices[ndim];
    int nnm1[ndim];
    for (int i = 0; i < ndim; i++) {
        nnm1[i] = n-1;
    }


#pragma omp parallel for reduction(+:total)
    for (int i = 0; i < npts; i++) {
        double xx[ndim];
        ind2sub(i, nnm1, indices);
        double jac = 1.0;
        for (int j = 0; j < ndim; j++) {
            double x = (indices[j]+1) * h;
            xx[j] = transform_cc(x);
            jac *= jacobian_cc(x);
        }
        total += jac*psi_fn(xx);
    }

    return total*std::pow(h, ndim);
}

int main()
{

    // Number of points in each dimension
    int n = 12;

    uint64_t npts = get_npts(ndim, n);

    // Compute for a single value of n
    std::cout << "n = " << n << " npts = " << npts << std::endl;
    double start = omp_get_wtime();
    double val = trapn_cc(n);
    double end = omp_get_wtime();
    std::cout << "val = " << std::setprecision(16) << val << std::endl;
    double elapsed = end-start;
    std::cout << " time = " << elapsed << "  eval rate = " << npts/elapsed << std::endl;


    // Scan different values of n
#if 0
    std::cout << "n   npts   val     elapsed time (s)    rate (evals/s) " << std::endl;
    for (int n = 10; n < 24; n++) {
      uint64_t npts = get_npts(ndim, n);
      double start = omp_get_wtime();
      double val = trapn_cc(n);
      double end = omp_get_wtime();
      double elapsed = end-start;
      std::cout << n << " "
                << npts <<  " "
                << val << " "
                << elapsed << " "
                << npts/elapsed << std::endl;
    }

#endif

}
